File: packages/libzaungast-native/src/sstable.rs
//! Minimal read-only LevelDB SSTable (.ldb) parser — a faithful port of
//! `packages/libzaungast/src/format/chromium/sstable.ts`. Read-only: opens the file, never writes,
//! never uses leveldb's DB::open (which would recover/compact = write). Corrupt/short blocks are
//! skipped and flagged `lossy`, exactly as the TS try/catch does — so a torn copy is never treated
//! as deletion truth downstream.
//! Format ref: https://github.com/google/leveldb/blob/main/doc/table_format.md

use crate::snappy;

const MAGIC: u64 = 0xdb4775248b80fb57; // full 64-bit table magic

/// CRC32C (Castagnoli) lookup table, built at compile time (mirrors sstable.ts's IIFE).
const CRC32C_TABLE: [u32; 256] = {
    let mut t = [0u32; 256];
    let mut n = 0usize;
    while n < 256 {
        let mut c = n as u32;
        let mut k = 0;
        while k < 8 {
            c = if c & 1 != 0 { 0x82f63b78 ^ (c >> 1) } else { c >> 1 };
            k += 1;
        }
        t[n] = c;
        n += 1;
    }
    t
};

fn crc32c(buf: &[u8], start: usize, end: usize) -> u32 {
    let mut c: u32 = 0xffffffff;
    let mut i = start;
    while i < end {
        c = CRC32C_TABLE[((c ^ buf[i] as u32) & 0xff) as usize] ^ (c >> 8);
        i += 1;
    }
    c ^ 0xffffffff
}

/// leveldb's mask: Mask(crc) = rotate_right(crc,15) + kMaskDelta (0xa282ead8). Unmask reverses it.
fn unmask_crc(m: u32) -> u32 {
    let x = m.wrapping_sub(0xa282ead8);
    (x >> 17) | (x << 15)
}

/// LEB128 varint (u64). `None` on truncation (bounds-safe — mirrors a TS RangeError → caught).
fn read_varint(buf: &[u8], off: usize) -> Option<(u64, usize)> {
    let mut result: u64 = 0;
    let mut shift: u32 = 0;
    let mut pos = off;
    loop {
        let byte = *buf.get(pos)?;
        pos += 1;
        result |= ((byte & 0x7f) as u64) << shift;
        if byte & 0x80 == 0 {
            break;
        }
        shift += 7;
        if shift >= 64 {
            return None;
        }
    }
    Some((result, pos))
}

/// A leveldb BlockHandle = (offset, size) as two varints.
fn read_block_handle(buf: &[u8], off: usize) -> Option<((usize, usize), usize)> {
    let (offset, pos) = read_varint(buf, off)?;
    let (size, pos) = read_varint(buf, pos)?;
    Some(((offset as usize, size as usize), pos))
}

struct BlockRead {
    data: Vec<u8>,
    crc_ok: Option<bool>,
}

/// Read a block's bytes: 5-byte trailer = 1 compression byte + 4-byte masked CRC. Type 0 = raw,
/// type 1 = snappy. `None` on any bounds/decompress failure or unsupported type (→ caller lossy).
/// (TS returns an uncompressed block as a zero-copy VIEW; we copy for now — byte-identical output,
/// optimize to Cow later.)
fn read_block(file: &[u8], offset: usize, size: usize, verify_crc: bool) -> Option<BlockRead> {
    let comp = *file.get(offset + size)?;
    let crc_ok = if verify_crc {
        let b = file.get(offset + size + 1..offset + size + 5)?;
        let stored = unmask_crc(u32::from_le_bytes([b[0], b[1], b[2], b[3]]));
        Some(crc32c(file, offset, offset + size + 1) == stored)
    } else {
        None
    };
    let contents = file.get(offset..offset + size)?;
    let data = match comp {
        0 => contents.to_vec(),
        1 => snappy::uncompress(contents).ok()?,
        _ => return None, // unsupported compression → treat as error (lossy), like the TS throw
    };
    Some(BlockRead { data, crc_ok })
}

/// Parse a data/index block's entries (restart-point prefix compression). `None` on corruption.
fn parse_block(data: &[u8]) -> Option<Vec<(Vec<u8>, Vec<u8>)>> {
    let n = data.len();
    if n < 4 {
        return None;
    }
    let num_restarts =
        u32::from_le_bytes([data[n - 4], data[n - 3], data[n - 2], data[n - 1]]) as usize;
    let restarts_start = n.checked_sub(4 + num_restarts.checked_mul(4)?)?;
    let mut pos = 0usize;
    let mut prev_key: Vec<u8> = Vec::new();
    let mut out: Vec<(Vec<u8>, Vec<u8>)> = Vec::new();
    while pos < restarts_start {
        let (shared, p) = read_varint(data, pos)?;
        pos = p;
        let (non_shared, p) = read_varint(data, pos)?;
        pos = p;
        let (value_len, p) = read_varint(data, pos)?;
        pos = p;
        let (shared, non_shared, value_len) =
            (shared as usize, non_shared as usize, value_len as usize);
        let key_start = pos;
        pos = pos.checked_add(non_shared)?;
        let value = data.get(pos..pos.checked_add(value_len)?)?.to_vec();
        pos += value_len;
        if shared > prev_key.len() {
            return None;
        }
        // key = shared prefix of prevKey + this entry's non-shared delta (single owned alloc).
        let mut key = Vec::with_capacity(shared + non_shared);
        key.extend_from_slice(&prev_key[..shared]);
        key.extend_from_slice(data.get(key_start..key_start + non_shared)?);
        prev_key = key.clone();
        out.push((key, value));
    }
    Some(out)
}

/// The result of reading a whole table: its (key, value) entries in table order, and whether any
/// block was skipped (lossy).
pub struct TableRead {
    pub entries: Vec<(Vec<u8>, Vec<u8>)>,
    pub lossy: bool,
}

fn lossy_empty() -> TableRead {
    TableRead { entries: Vec::new(), lossy: true }
}

/// Read every entry from an .ldb table file. Verifies the index-block CRC (a torn copy loses the
/// tail → fails here → lossy); per-data-block CRC skipped for speed (matches the TS reader).
pub fn read_table(path: &str) -> std::io::Result<TableRead> {
    let file = std::fs::read(path)?;
    let len = file.len();
    if len < 48 {
        return Ok(lossy_empty());
    }
    let footer_off = len - 48;
    let magic = u64::from_le_bytes(file[len - 8..len].try_into().unwrap());
    if magic != MAGIC {
        eprintln!("WARNING: bad magic {:x} (expected {:x})", magic, MAGIC);
    }
    // footer: metaindex handle (skipped) then index handle
    let (_, p) = match read_block_handle(&file, footer_off) {
        Some(x) => x,
        None => return Ok(lossy_empty()),
    };
    let (index_handle, _) = match read_block_handle(&file, p) {
        Some(x) => x,
        None => return Ok(lossy_empty()),
    };
    let index_block = match read_block(&file, index_handle.0, index_handle.1, true) {
        Some(b) => b,
        None => return Ok(lossy_empty()),
    };
    if index_block.crc_ok == Some(false) {
        return Ok(lossy_empty());
    }
    let index_entries = match parse_block(&index_block.data) {
        Some(e) => e,
        None => return Ok(lossy_empty()),
    };
    let mut entries: Vec<(Vec<u8>, Vec<u8>)> = Vec::new();
    let mut lossy = false;
    for (_ikey, ivalue) in &index_entries {
        let handle = match read_block_handle(ivalue, 0) {
            Some((h, _)) => h,
            None => {
                lossy = true;
                continue;
            }
        };
        match read_block(&file, handle.0, handle.1, false) {
            Some(b) => match parse_block(&b.data) {
                Some(es) => entries.extend(es),
                None => lossy = true,
            },
            None => lossy = true,
        }
    }
    Ok(TableRead { entries, lossy })
}

/// (count, crc32c) digest of a table's entries — the cross-language equality oracle. Feeds, per
/// entry in order: keyLen (u32 LE), key bytes, valLen (u32 LE), value bytes. The TS harness computes
/// the identical digest; matching (count, crc) ⇒ byte-identical reads.
pub fn entries_digest(entries: &[(Vec<u8>, Vec<u8>)]) -> (usize, u32) {
    let mut c: u32 = 0xffffffff;
    let mut feed = |b: u8| c = CRC32C_TABLE[((c ^ b as u32) & 0xff) as usize] ^ (c >> 8);
    for (k, v) in entries {
        for b in (k.len() as u32).to_le_bytes() {
            feed(b);
        }
        for &b in k {
            feed(b);
        }
        for b in (v.len() as u32).to_le_bytes() {
            feed(b);
        }
        for &b in v {
            feed(b);
        }
    }
    (entries.len(), c ^ 0xffffffff)
}

File: packages/libzaungast/src/format/chromium/sstable.ts
// Minimal read-only LevelDB SSTable (.ldb) parser.
// Format ref: https://github.com/google/leveldb/blob/main/doc/table_format.md
import fs from 'node:fs';
import * as Snappy from './snappy.js';
import type { BlockHandle, BlockReadResult, TableEntry, TableReadResult } from '../types.js';

const MAGIC_HIGH = 0xdb4775248b80fb57n; // full 64-bit magic

function readVarint(buf: Buffer, off: number): [bigint, number] {
  let result = 0n;
  let shift = 0n;
  let pos = off;
  while (true) {
    const byte = buf[pos++];
    result |= BigInt(byte & 0x7f) << shift;
    if ((byte & 0x80) === 0) break;
    shift += 7n;
  }
  return [result, pos];
}
function readVarintNum(buf: Buffer, off: number): [number, number] {
  const [v, pos] = readVarint(buf, off);
  return [Number(v), pos];
}

function readBlockHandle(buf: Buffer, off: number): [BlockHandle, number] {
  let offset: number, size: number, pos: number;
  [offset, pos] = readVarintNum(buf, off);
  [size, pos] = readVarintNum(buf, pos);
  return [{ offset, size }, pos];
}

// CRC32C (Castagnoli) with leveldb's masking, for block-integrity verification.
const CRC32C_TABLE = (() => {
  const t = new Uint32Array(256);
  for (let n = 0; n < 256; n++) {
    let c = n;
    for (let k = 0; k < 8; k++) c = c & 1 ? 0x82f63b78 ^ (c >>> 1) : c >>> 1;
    t[n] = c >>> 0;
  }
  return t;
})();
export function crc32c(buf: Buffer, start: number, end: number): number {
  let c = 0xffffffff;
  for (let i = start; i < end; i++) c = CRC32C_TABLE[(c ^ buf[i]) & 0xff] ^ (c >>> 8);
  return (c ^ 0xffffffff) >>> 0;
}
// leveldb's masking: Mask(crc) = rotate_right(crc,15) + kMaskDelta; kMaskDelta = 0xa282ead8.
export function unmaskCrc(m: number): number {
  const x = (m - 0xa282ead8) >>> 0;
  return ((x >>> 17) | (x << 15)) >>> 0;
}

// Read a block's raw bytes from the in-memory file buffer, handling the 5-byte trailer +
// compression. Uncompressed blocks are returned as a *view* into `file` (no copy) — the caller
// must treat block/entry buffers as read-only (nothing in the reader mutates decoded buffers).
function readBlock(file: Buffer, handle: BlockHandle, verifyCrc = false): BlockReadResult {
  const { offset, size } = handle;
  const compressionType = file[offset + size];
  let crcOk: boolean | null = null;
  if (verifyCrc) {
    const stored = unmaskCrc(file.readUInt32LE(offset + size + 1));
    crcOk = crc32c(file, offset, offset + size + 1) === stored;
  }
  const contents = file.subarray(offset, offset + size);
  if (compressionType === 0) return { data: contents, compressionType, crcOk };
  if (compressionType === 1) return { data: Snappy.uncompress(contents), compressionType, crcOk };
  throw new Error(`Unsupported compression type ${compressionType}`);
}

export { readBlock, readBlockHandle };

// Parse the entries of a data/index block (with restart-point prefix compression).
export function* parseBlock(data: Buffer): Generator<TableEntry> {
  const n = data.length;
  // last 4 bytes = num_restarts
  const numRestarts = data.readUInt32LE(n - 4);
  const restartsStart = n - 4 - numRestarts * 4;
  let pos = 0;
  let prevKey = Buffer.alloc(0);
  while (pos < restartsStart) {
    let shared: number, nonShared: number, valueLen: number;
    [shared, pos] = readVarintNum(data, pos);
    [nonShared, pos] = readVarintNum(data, pos);
    [valueLen, pos] = readVarintNum(data, pos);
    const keyStart = pos;
    pos += nonShared;
    const value = data.subarray(pos, pos + valueLen); // view into the block
    pos += valueLen;
    // Build the key with a single owned allocation (shared prefix + this entry's delta) instead of
    // Buffer.concat's array literal + intermediate subarray.
    const key = Buffer.allocUnsafe(shared + nonShared);
    prevKey.copy(key, 0, 0, shared);
    data.copy(key, shared, keyStart, keyStart + nonShared);
    prevKey = key;
    yield [key, value];
  }
}

export function readTable(path: string): TableReadResult {
  // Read the whole (tmp-copy) file once: one syscall, no per-block zeroed Buffer.alloc + readSync.
  // Blocks/values become views into this buffer; the entries returned pin it for their lifetime.
  const file = fs.readFileSync(path);
  const fileSize = file.length;
  const footerOff = fileSize - 48;
  const magic = file.readBigUInt64LE(fileSize - 8);
  if (magic !== MAGIC_HIGH) {
    console.error(`WARNING: bad magic ${magic.toString(16)} (expected ${MAGIC_HIGH.toString(16)})`);
  }
  // metaindex handle (skipped), then index handle (varints at the start of the 48-byte footer)
  const [, p] = readBlockHandle(file, footerOff);
  const [indexHandle] = readBlockHandle(file, p);

  // Verify CRCs: a copy taken while a table was being written could be torn. If the index
  // block itself is corrupt we can't trust the table — skip it. Corrupt data blocks are
  // skipped individually.
  // Verifying the INDEX block CRC (plus the footer magic check above) is enough to detect a
  // torn/truncated .ldb copy — a partial copy loses the tail (footer/index) and fails here.
  // Per-data-block CRC is skipped for speed (halves read time; the WAL is still fully CRC'd).
  // `lossy` = true if any block was skipped, so callers can refuse to treat a partial read as
  // deletion truth (an incremental deletion reconcile MUST NOT run on a lossy load).
  const { data: indexData, crcOk: idxOk } = readBlock(file, indexHandle, true);
  if (idxOk === false) return { entries: [], lossy: true };

  const entries: TableEntry[] = [];
  let lossy = false;
  for (const [, ivalue] of parseBlock(indexData)) {
    // index value = BlockHandle of a data block
    const [handle] = readBlockHandle(ivalue, 0);
    try {
      const block = readBlock(file, handle, false);
      // key is owned (parseBlock allocates it); value is a view into the block (into `file` for
      // uncompressed blocks, or the per-block snappy output). No copy here.
      for (const [k, v] of parseBlock(block.data)) entries.push([k, v]);
    } catch {
      lossy = true;
      continue;
    } // corrupt/short block → skip, mark lossy
  }
  return { entries, lossy };
}
